packages/shared/src/utils/collectionLayout.ts

```diff
@@ -0,0 +1,114 @@
+import { graphql, readInlineData } from 'relay-runtime';
+
+import { collectionLayoutParseFragment$key } from '~/generated/collectionLayoutParseFragment.graphql';
+
+import { DEFAULT_COLUMNS } from '../constants/layout';
+import { removeNullValues } from '../relay/removeNullValues';
+import { generate12DigitId } from './generate12DigitId';
+
+export type WhitespaceBlock = {
+  id: string;
+  whitespace: 'whitespace';
+};
+
+export type Section<T> = {
+  id: string;
+  items: ReadonlyArray<T | WhitespaceBlock>;
+  columns: number;
+};
+export type CollectionWithLayout<T> = Array<Section<T>>;
+
+export type CollectionLayout = {
+  sections: Array<number>;
+  sectionLayout: Array<{ whitespace: ReadonlyArray<number | null> | null; columns: number | null }>;
+};
+
+export function parseCollectionLayoutGraphql<T>(
+  tokens: T[],
```

Comment:
Move most of the editor logic from **web** to **shared**. Still have some left and will refactor it later to be more generic. Right now, its really tied to the **types**

Reply:
amazing